### ISSUE TITLE
Make controller available to serialize/deserialize QPs

### ISSUE THREAD

**Issue by mwpastore** [offirgolan/ember-parachute#27]
I don't know if it is possible, but it would be incredibly useful for the serialize and deserialize functions to have access to the object the query params are mixed into. For example, consider a query param that represents an array index. The deserialize function should be able to sanity check that index to make sure it's in range. 

```javascript
const myQueryParams = new QueryParams({
  foo: {
    deserialize(value) {
      const length = this.get('bar.length'); // TypeError: this.get is not a function
      // ..
    }
  }
});

export default Ember.Controller.extend(myQueryParams.Mixin, {
  bar: []
});
```

**Comment by offirgolan** [comment]
@mwpastore although this would be nice to have, query param serialization / deserialization can happen before the controller is even created so I dont believe this would be currently possible.

**Comment by poteto** [comment]
this might be possible... I'll whip something up when I wake up

**Comment by poteto** [comment]
I had a bit of a think about this, and it seems to me like the issue is really about *where* this kind of validation of query params should live. In my mind, the serialize / deserialize functions are purely for presentation of the query parameter in the URL only. Validation or other business logic should happen before the value is set on the controller. In your case, you should be validating the item's index in the array **before** setting the value.

Does that make sense? @mwpastore

**Comment by mwpastore** [comment]
@poteto Thanks for your thoughts and work on this. I understand what you're saying and basically agree, but if a user first browsers to the site with e.g. `?index=3`, ideally there would be a way to sanity check that index against the array it's selecting into. Perhaps that's an argument for having a separate setter function in the controller instead of enhancing the deserialize QP property. Either way it would be good to figure out and document this use case.

**Comment by offirgolan** [comment]
@mwpastore the idea behind the serialize/deserialize is not to do any business logic but only to format how it should look in the URI. The business logic should be done before setting the query parameter in the first place.
